### PR TITLE
Follow  Unfollow User

### PR DESCRIPTION
**Pull request by SSalekin**

### CODE REVIEW

File: app/controllers/relationships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class RelationshipsController < ApplicationController
+  def create
+    @user = User.find(params[:followed_id])
```
Comment (abdullahallmehedi) [inline app/controllers/relationships_controller.rb]: ()

File: app/models/user.rb [hunk 2]
```diff
@@ -22,4 +22,16 @@ class User < ActiveRecord::Base
   mount_uploader :image, UserImageUploader
   validates :image, file_size: {less_than: 3.megabytes},
     file_content_type: {allow: ["image/jpeg", "image/png"]}
-end
\ No newline at end of file
+
+  def follow other_user
+    active_relationships.create(followed_id: other_user.id)
```
Comment (abdullahallmehedi) [inline app/models/user.rb]: ()

File: app/models/user.rb [hunk 3]
```diff
@@ -22,4 +22,16 @@ class User < ActiveRecord::Base
...
+  end
+
+  def unfollow other_user
+    active_relationships.find_by(followed_id: other_user.id).destroy
+  end
+
+  def following? other_user
+    following.include?(other_user)
```
Comment (abdullahallmehedi) [inline app/models/user.rb]: ()

File: app/views/users/_follow_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<% unless current_user == @user %>
+  <div id="follow_form" class="pull-right">
+    <% if current_user.following?(@user) %>
```
Comment (abdullahallmehedi) [inline app/views/users/_follow_form.html.erb]: ()

File: app/views/users/_unfollow.html.erb [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<%= form_for(current_user.active_relationships.find_by(followed_id: @user.id),
+  html: { method: :delete }, remote: true) do |f| %>
+  <%= f.submit "Unfollow", class: "btn" %>
```
Comment (abdullahallmehedi) [inline app/views/users/_unfollow.html.erb]: i18n

File: config/locales/devise.en.yml [hunk 6]
```diff
@@ -58,5 +62,5 @@ en:
       not_found: not found
       not_locked: was not locked
       not_saved:
-        one: "1 error prohibited this %{resource} from being saved:"
-        other: "%{count} errors prohibited this %{resource} from being saved:"
+        one: '1 error prohibited this %{resource} from being saved:'
```
Comment (abdullahallmehedi) [inline config/locales/devise.en.yml]: ""

File: config/locales/devise.en.yml [hunk 7]
```diff
@@ -58,5 +62,5 @@ en:
...
+        other: '%{count} errors prohibited this %{resource} from being saved:'
```
Comment (abdullahallmehedi) [inline config/locales/devise.en.yml]: ""

**Comment by noor-framgia** [comment]
ok

**Comment by Rownaknira** [comment]
ok

File: app/views/users/_favorite_books.html.erb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+<div class="panel-body">
+  <div class="list-group">
+    <% Book.favourited_by(@user).each do |book| %>
```
Comment (dieunb) [inline app/views/users/_favorite_books.html.erb]: assign `Book.favourited_by(@user)` to a variable and move it on controller

File: config/routes.rb [hunk 9]
```diff
@@ -12,8 +12,13 @@
     get "contact"=> "static_pages#contact"
 
     resources :requests, only: [:new, :create, :index, :destroy]
-    resources :users, only: [:index, :show, :edit, :update]
+    resources :users, only: [:index, :show, :edit, :update] do
+      member do
+        get :following, :followers
```
Comment (dieunb) [inline config/routes.rb]: where are functions?

Comment (SSalekin) [inline config/routes.rb]: Anh, I didn't understand your question. Did you mean where we used the following, follower functions?
I call these from _followers.html.erb and _followings.html.erb partial.

File: app/views/users/_followers.html.erb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+<div class="followers well" id="followers_div">
+    <button type="button" class="close" id="hide-followers">&times;</button>
```
Comment (dieunb) [inline app/views/users/_followers.html.erb]: indent

File: app/views/users/_followings.html.erb [hunk 11]
```diff
@@ -0,0 +1,14 @@
+<div class="following well" id="following_div">
+    <button type="button" class="close" id="hide-following">&times;</button>
```
Comment (dieunb) [inline app/views/users/_followings.html.erb]: indent

File: config/routes.rb [hunk 12]
```diff
@@ -12,8 +12,14 @@
     get "contact"=> "static_pages#contact"
 
     resources :requests, only: [:new, :create, :index, :destroy]
-    resources :users, only: [:index, :show, :edit, :update]
-    resources :books, only: [:index, :show]
+
+    resources :users, only: [:index, :show, :edit, :update] do
+      member do
+        get :following, :followers
```
Comment (dieunb) [inline config/routes.rb]: where do you define it?

Comment (SSalekin) [inline config/routes.rb]: I wanted to add these when implementing Decorator.

**Pull request merged**